File: Data Structures & Algorithms/interleaving-string/submission-0.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        if (s1.size() + s2.size() != s3.size()) {
            return false;
        }

        int m = s1.size();
        int n = s2.size();

        vector<vector<bool>> dp(m + 1, vector<bool> (n +1, false));
        dp[m][n] = true;

        for (int i = m; i >= 0; i--) {
            for (int j = n; j >= 0; j--){
                if (i <  m && s1[i] == s3[i+j] && dp[i+1][j]) dp[i][j] = true;
                else if (j <  n && s2[j] == s3[i+j] && dp[i][j+1]) dp[i][j] = true;
            }
        }

        return dp[0][0];
    }
};

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows, cols = len(matrix), len(matrix[0])
        max_seq = 0

        dp = {}
        
        def check(r: int, c: int) -> bool:
            if (0 <= r < rows) and (0 <= c < cols):
                return True
            return False

        dirs = [(0,1), (1,0), (-1,0), (0,-1)]
        def dfs(sr: int, sc: int):
            if (sr,sc) in dp:
                return dp[(sr,sc)]
            res = 1
            for d1, d2 in dirs:
                nr, nc = sr + d1, sc + d2
                if not check(nr, nc):
                    continue
                if matrix[nr][nc] > matrix[sr][sc]:
                    res = max(res, 1 + dfs(nr,nc))

            dp[(sr,sc)] = res
            return res

                
        for i in range(rows):
            for j in range(cols):
                max_seq = max(max_seq, dfs(i, j))
       

        return max_seq